x/mongo/driver/topology/topology_options.go

```diff
@@ -7,451 +7,370 @@
+	// SocketTimeout
+	if co.SocketTimeout != nil {
+		connOpts = append(
+			connOpts,
+			WithReadTimeout(func(time.Duration) time.Duration { return *co.SocketTimeout }),
+			WithWriteTimeout(func(time.Duration) time.Duration { return *co.SocketTimeout }),
+		)
+	}
+	// TLSConfig
+	if co.TLSConfig != nil {
+		connOpts = append(connOpts, WithTLSConfig(
+			func(*tls.Config) *tls.Config {
+				return co.TLSConfig
+			},
+		))
 	}
 
-	return certBlock.Bytes, nil
-}
+	// OCSP cache
+	ocspCache := ocsp.NewCache()
+	connOpts = append(
+		connOpts,
+		WithOCSPCache(func(ocsp.Cache) ocsp.Cache { return ocspCache }),
+	)
+
+	// Disable communication with external OCSP responders.
+	if co.DisableOCSPEndpointCheck != nil {
+		connOpts = append(
+			connOpts,
+			WithDisableOCSPEndpointCheck(func(bool) bool { return *co.DisableOCSPEndpointCheck }),
+		)
+	}
 
-// addClientCertFromFile adds a client certificate to the configuration given a path to the
-// containing file and returns the certificate's subject name.
-func addClientCertFromFile(cfg *tls.Config, clientFile, keyPasswd string) (string, error) {
-	data, err := ioutil.ReadFile(clientFile)
-	if err != nil {
-		return "", err
+	// LoadBalanced
+	if co.LoadBalanced != nil {
+		cfgp.LoadBalanced = *co.LoadBalanced
+
+		serverOpts = append(
+			serverOpts,
+			WithServerLoadBalanced(func(bool) bool { return *co.LoadBalanced }),
+		)
+		connOpts = append(
+			connOpts,
+			WithConnectionLoadBalanced(func(bool) bool { return *co.LoadBalanced }),
+		)
 	}
 
-	var currentBlock *pem.Block
-	var certBlock, certDecodedBlock, keyBlock []byte
+	serverOpts = append(
+		serverOpts,
+		WithClock(func(*session.ClusterClock) *session.ClusterClock { return clusterClock }),
+		WithConnectionOptions(func(...ConnectionOption) []ConnectionOption { return connOpts }))
 
-	remaining := data
-	start := 0
-	for {
-		currentBlock, remaining = pem.Decode(remaining)
-		if currentBlock == nil {
-			break
-		}
+	cfgp.ServerOpts = serverOpts
 
-		if currentBlock.Type == "CERTIFICATE" {
-			certBlock = data[start : len(data)-len(remaining)]
-			certDecodedBlock = currentBlock.Bytes
-			start += len(certBlock)
-		} else if strings.HasSuffix(currentBlock.Type, "PRIVATE KEY") {
-			if keyPasswd != "" && x509.IsEncryptedPEMBlock(currentBlock) {
-				var encoded bytes.Buffer
-				buf, err := x509.DecryptPEMBlock(currentBlock, []byte(keyPasswd))
-				if err != nil {
-					return "", err
-				}
-
-				pem.Encode(&encoded, &pem.Block{Type: currentBlock.Type, Bytes: buf})
-				keyBlock = encoded.Bytes()
-				start = len(data) - len(remaining)
-			} else {
-				keyBlock = data[start : len(data)-len(remaining)]
-				start += len(keyBlock)
-			}
+	// Deployment
+	if co.Deployment != nil {
+		// topology options: WithSeedlist, WithURI, WithSRVServiceName, WithSRVMaxHosts, and WithServerOptions
+		// server options: WithClock and WithConnectionOptions + default maxPoolSize
+		if len(serverOpts) > 2+defaultOptions || hasOptionalData(*co) {
```

Comment:
This change is of note. The original design looked at the length of the topology options slice to see if we were over assigning a number of optional topology options. That isn't possible anymore since we have deprecated the options pattern. So, instead, we use the `hasOptionalData` function to see if the underlying client options have more than the minimum data.

Reply:
This seems like a fine alternative to me. The comment above is now (and was) a little confusing though. Can we update it to explain what exactly we're checking for?